File: backend-repo/examples/his-integration/java/MrrArchiveTicketClient.java
package example.mrr;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.Duration;
import java.time.Instant;
import java.util.HexFormat;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

/**
 * HIS/EMR 服务端申请 MRR 外部影像档案袋一次性票据。
 *
 * <p>HMAC Secret 只能保存在 HIS/EMR 服务端。浏览器、Vue/React、WPF/WinForms
 * 客户端都不应直接持有 Secret。</p>
 */
public final class MrrArchiveTicketClient {

    private static final String TICKET_PATH = "/api/v1/integration/archive/tickets";
    private static final String HMAC_ALGORITHM = "HmacSHA256";

    private final URI baseUri;
    private final String clientId;
    private final String secret;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;

    public MrrArchiveTicketClient(String baseUrl, String clientId, String secret) {
        this(
                baseUrl,
                clientId,
                secret,
                HttpClient.newBuilder()
                        .connectTimeout(Duration.ofSeconds(10))
                        .build(),
                new ObjectMapper()
        );
    }

    public MrrArchiveTicketClient(
            String baseUrl,
            String clientId,
            String secret,
            HttpClient httpClient,
            ObjectMapper objectMapper
    ) {
        this.baseUri = URI.create(requireNonBlank(baseUrl, "baseUrl").replaceAll("/+$", ""));
        this.clientId = requireNonBlank(clientId, "clientId");
        this.secret = requireNonBlank(secret, "secret");
        this.httpClient = Objects.requireNonNull(httpClient, "httpClient");
        this.objectMapper = Objects.requireNonNull(objectMapper, "objectMapper").copy()
                .setSerializationInclusion(JsonInclude.Include.NON_EMPTY);
    }

    public TicketResult createTicket(TicketRequest payload) throws Exception {
        Objects.requireNonNull(payload, "payload");
        if (payload.externalUserId() == null || payload.externalUserId().isBlank()) {
            throw new IllegalArgumentException("externalUserId 不能为空");
        }

        // 必须先序列化一次，并对同一份字节同时计算 SHA-256 和发送 HTTP 请求。
        byte[] rawBody = objectMapper.writeValueAsBytes(payload);
        String timestamp = String.valueOf(Instant.now().getEpochSecond());
        String nonce = UUID.randomUUID().toString();
        String bodyHash = HexFormat.of().formatHex(
                MessageDigest.getInstance("SHA-256").digest(rawBody)
        );
        String canonicalText = "POST\n" + TICKET_PATH + "\n"
                + timestamp + "\n" + nonce + "\n" + bodyHash;
        String signature = hmacSha256Hex(secret, canonicalText);

        HttpRequest request = HttpRequest.newBuilder(baseUri.resolve(TICKET_PATH))
                .timeout(Duration.ofSeconds(30))
                .header("Content-Type", "application/json; charset=utf-8")
                .header("X-MRR-Client-Id", clientId)
                .header("X-MRR-Timestamp", timestamp)
                .header("X-MRR-Nonce", nonce)
                .header("X-MRR-Signature", signature)
                .POST(HttpRequest.BodyPublishers.ofByteArray(rawBody))
                .build();

        HttpResponse<String> response = httpClient.send(
                request,
                HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8)
        );
        JsonNode root = parseJson(response.body(), response.statusCode());
        int businessCode = root.path("code").asInt(response.statusCode());
        String message = root.path("message").asText("");
        if (response.statusCode() < 200 || response.statusCode() >= 300 || businessCode != 200) {
            throw new MrrIntegrationException(response.statusCode(), businessCode, message, response.body());
        }

        JsonNode data = root.path("data");
        String ticket = data.path("ticket").asText("");
        String launchUrl = data.path("launchUrl").asText("");
        if (ticket.isBlank() || launchUrl.isBlank()) {
            throw new MrrIntegrationException(
                    response.statusCode(), businessCode,
                    "MRR 响应缺少 ticket 或 launchUrl", response.body()
            );
        }
        return new TicketResult(
                ticket,
                launchUrl,
                data.path("expiresIn").asInt(),
                data.path("archiveCount").asInt()
        );
    }

    public String createLaunchUrl(TicketRequest payload) throws Exception {
        return createTicket(payload).launchUrl();
    }

    private JsonNode parseJson(String body, int httpStatus) throws Exception {
        try {
            return objectMapper.readTree(body);
        } catch (Exception exception) {
            throw new MrrIntegrationException(
                    httpStatus, httpStatus,
                    "MRR 返回的内容不是有效 JSON", body
            );
        }
    }

    private static String hmacSha256Hex(String secret, String canonicalText) throws Exception {
        Mac mac = Mac.getInstance(HMAC_ALGORITHM);
        mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), HMAC_ALGORITHM));
        return HexFormat.of().formatHex(
                mac.doFinal(canonicalText.getBytes(StandardCharsets.UTF_8))
        );
    }

    private static String requireNonBlank(String value, String name) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException(name + " 不能为空");
        }
        return value.trim();
    }

    public record ArchiveSelector(String bah, String sjh) {
    }

    public record TicketRequest(
            String externalUserId,
            String idCard,
            String bah,
            String sjh,
            List<String> bahs,
            List<String> sjhs,
            List<ArchiveSelector> archives,
            boolean allowDownload
    ) {
        public static TicketRequest exactArchive(
                String externalUserId,
                String bah,
                String sjh,
                boolean allowDownload
        ) {
            return new TicketRequest(
                    externalUserId,
                    null,
                    null,
                    null,
                    List.of(),
                    List.of(),
                    List.of(new ArchiveSelector(bah, sjh)),
                    allowDownload
            );
        }
    }

    public record TicketResult(
            String ticket,
            String launchUrl,
            int expiresIn,
            int archiveCount
    ) {
    }

    public static final class MrrIntegrationException extends RuntimeException {
        private final int httpStatus;
        private final int businessCode;
        private final String responseBody;

        public MrrIntegrationException(
                int httpStatus,
                int businessCode,
                String message,
                String responseBody
        ) {
            super("MRR 调用失败：HTTP " + httpStatus
                    + ", code=" + businessCode
                    + ", message=" + message);
            this.httpStatus = httpStatus;
            this.businessCode = businessCode;
            this.responseBody = responseBody;
        }

        public int getHttpStatus() {
            return httpStatus;
        }

        public int getBusinessCode() {
            return businessCode;
        }

        public String getResponseBody() {
            return responseBody;
        }
    }
}

File: backend-repo/examples/his-integration/python/mrr_archive_ticket_client.py
#!/usr/bin/env python3
"""HIS/EMR 服务端申请 MRR 外部影像档案袋一次性票据。"""

from __future__ import annotations

import hashlib
import hmac
import json
import os
import sys
import time
import urllib.error
import urllib.request
import uuid
from dataclasses import dataclass
from typing import Any

TICKET_PATH = "/api/v1/integration/archive/tickets"


class MrrIntegrationError(RuntimeError):
    def __init__(
        self,
        http_status: int,
        business_code: int,
        message: str,
        response_body: str,
    ) -> None:
        super().__init__(
            f"MRR 调用失败：HTTP {http_status}, "
            f"code={business_code}, message={message}"
        )
        self.http_status = http_status
        self.business_code = business_code
        self.response_body = response_body


@dataclass(frozen=True)
class TicketResult:
    ticket: str
    launch_url: str
    expires_in: int
    archive_count: int


class MrrArchiveTicketClient:
    def __init__(self, base_url: str, client_id: str, secret: str) -> None:
        self.base_url = self._require(base_url, "base_url").rstrip("/")
        self.client_id = self._require(client_id, "client_id")
        self.secret = self._require(secret, "secret")

    def create_ticket(self, payload: dict[str, Any], timeout: int = 30) -> TicketResult:
        external_user_id = str(payload.get("externalUserId") or "").strip()
        if not external_user_id:
            raise ValueError("externalUserId 不能为空")

        # 只序列化一次：签名和 HTTP 请求必须使用完全相同的 UTF-8 字节。
        raw_body = json.dumps(
            payload,
            ensure_ascii=False,
            separators=(",", ":"),
        ).encode("utf-8")
        timestamp = str(int(time.time()))
        nonce = str(uuid.uuid4())
        body_hash = hashlib.sha256(raw_body).hexdigest()
        canonical_text = (
            f"POST\n{TICKET_PATH}\n{timestamp}\n{nonce}\n{body_hash}"
        )
        signature = hmac.new(
            self.secret.encode("utf-8"),
            canonical_text.encode("utf-8"),
            hashlib.sha256,
        ).hexdigest()

        request = urllib.request.Request(
            self.base_url + TICKET_PATH,
            data=raw_body,
            method="POST",
            headers={
                "Content-Type": "application/json; charset=utf-8",
                "X-MRR-Client-Id": self.client_id,
                "X-MRR-Timestamp": timestamp,
                "X-MRR-Nonce": nonce,
                "X-MRR-Signature": signature,
            },
        )

        try:
            with urllib.request.urlopen(request, timeout=timeout) as response:
                http_status = response.status
                response_body = response.read().decode("utf-8")
        except urllib.error.HTTPError as error:
            http_status = error.code
            response_body = error.read().decode("utf-8", errors="replace")
        except urllib.error.URLError as error:
            raise RuntimeError(f"无法连接 MRR：{error.reason}") from error

        try:
            api_response = json.loads(response_body)
        except json.JSONDecodeError as error:
            raise MrrIntegrationError(
                http_status,
                http_status,
                "MRR 返回的内容不是有效 JSON",
                response_body,
            ) from error

        business_code = int(api_response.get("code", http_status))
        if not 200 <= http_status < 300 or business_code != 200:
            raise MrrIntegrationError(
                http_status,
                business_code,
                str(api_response.get("message") or "MRR 调用失败"),
                response_body,
            )

        data = api_response.get("data") or {}
        ticket = str(data.get("ticket") or "")
        launch_url = str(data.get("launchUrl") or "")
        if not ticket or not launch_url:
            raise MrrIntegrationError(
                http_status,
                business_code,
                "MRR 响应缺少 ticket 或 launchUrl",
                response_body,
            )

        return TicketResult(
            ticket=ticket,
            launch_url=launch_url,
            expires_in=int(data.get("expiresIn") or 0),
            archive_count=int(data.get("archiveCount") or 0),
        )

    def create_launch_url(self, payload: dict[str, Any], timeout: int = 30) -> str:
        return self.create_ticket(payload, timeout).launch_url

    @staticmethod
    def _require(value: str, name: str) -> str:
        normalized = (value or "").strip()
        if not normalized:
            raise ValueError(f"{name} 不能为空")
        return normalized


def require_environment(name: str) -> str:
    value = os.getenv(name, "").strip()
    if not value:
        raise RuntimeError(f"缺少环境变量 {name}")
    return value


def main() -> int:
    if len(sys.argv) < 4:
        print(
            "用法: python mrr_archive_ticket_client.py "
            "<externalUserId> <bah> <sjh>",
            file=sys.stderr,
        )
        print(
            "需要环境变量: MRR_BASE_URL, MRR_CLIENT_ID, MRR_HMAC_SECRET",
            file=sys.stderr,
        )
        return 2

    client = MrrArchiveTicketClient(
        require_environment("MRR_BASE_URL"),
        require_environment("MRR_CLIENT_ID"),
        require_environment("MRR_HMAC_SECRET"),
    )
    payload = {
        "externalUserId": sys.argv[1],
        "archives": [
            {
                "bah": sys.argv[2],
                "sjh": sys.argv[3],
            }
        ],
        "allowDownload": False,
    }

    try:
        result = client.create_ticket(payload)
    except MrrIntegrationError as error:
        print(str(error), file=sys.stderr)
        print(error.response_body, file=sys.stderr)
        return 1

    print(f"archiveCount={result.archive_count}")
    print(f"expiresIn={result.expires_in}")
    print(result.launch_url)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
